### DISCUSSION TITLE
mconfig: `uname` doen't really work on OpenWrt. Any ideas?

### DISCUSSION THREAD

**Discussion by IngwiePhoenix** [davmac314/dinit#154] [Q&A] [answered]
So on most standard OpenWrts, you aren't exactly supposed to run a secondary init system. However, some far more powerful devices could totally profit from it, especially since you can `opkg install apk`, add the Apline repos, and basically have a full-blown secondary package manager. In my case, I do that to separate "router stuff" from "all the other stuff". Jellyfin, TVHeadend and many other tools are installed this way. Even Mesa, as the SBC I am using actually has an integrated GPU, which I can use to slightly lighten the load when running comskip and other tools. Its neat.

However, for other tools like Caddy and Clash, it is much more beneficial to just run a small secondary init for "user stuff". That is what I use dinit for.

But when building from source, `configs/mconfig.$$UNAME` is used, which even on OpenWrt resolves to `Linux`. But: OpenWrt uses the musl libc, and thus needs `-D_GLIBCXX_USE_CXX11_ABI=0` set.

Since I am not 100% sure how to best add this config, I wanted to ask what best to do.

Here is the `mconfig` I use:

```
# Installation path options.

SBINDIR=/sbin
MANDIR=/usr/share/man
SYSCONTROLSOCKET=/run/dinitctl

# Can be obtained from opkg: opkg install gcc
CXX=g++
# Default options, but MUSL is not GLIBC.
CXXOPTS=-D_GLIBCXX_USE_CXX11_ABI=0 -std=c++11 -Os -Wall -flto -fno-rtti -fno-plt
LDFLAGS=
# procd has that, not needed.
BUILD_SHUTDOWN=no
SANITIZEOPTS=

# Feature settings
# This causes error messages about not finding the root one.
# But since it's also just secondary, it can be off.
SUPPORT_CGROUPS=0
```

Thanks! =)

**Comment by mobin-2008** [comment]
Hi.
I converted your issue to discussion; because it's better place to talk about your issue (For example you can set a reply as answer).

**Comment by davmac314** [comment, marked as answer]
> But: OpenWrt uses the musl libc, and thus needs -D_GLIBCXX_USE_CXX11_ABI=0 set.

`_GLIBCXX_USE_CXX11_ABI` controls the ABI used within libstdc++, which is normally built as part of GCC. So, unless this is a misunderstanding on my part (in which case, please explain!), whether or not musl is your libc has no bearing on whether you should compile with `_GLIBCXX_USE_CXX11_ABI` set to a particular value. Despite the name of the macro, it has nothing to do with Glibc.

You should use `_GLIBCXX_USE_CXX11_ABI=0` if you are building with GCC 5 or 6, and `_GLIBCXX_USE_CXX_ABI=1` if building with GCC 7. However even for GCC 6 and 7 this is no longer required with the current point releases, since I note the fix for was backported even to GCC 6.

In terms of adding new config for OpenWRT, I suggest either:
- add a file in `configs` (eg `mconfig.OpenWRT`) but accept that it won't be automatically selected by the current build system; people can use by manually copying or linking it to `mconfig` before running `make`
- and/or, adjust the `mconfig.Linux.sh` shell script which is used for configuration on Linux. You are free to use whatever mechanism you like then to distinguish whether the build is for OpenWRT (so long as you don't break non-OpenWRT builds of course!).

**Reply by IngwiePhoenix** [reply]
> ABI used within libstdc++, which is normally built as part of GCC.

Not quite. :) The system's standard C library - I'll just use the term "libc" - is also the basis for the C++ standard library. The GNU C Library - glibc - also has a deriviation. When you build GCC, look for a folder of `GLIBCXX_$version`. However, it's not the only version out there.

This answer explains it quite quickly: https://stackoverflow.com/a/73628686/2423150

Now since OpenWrt uses the MUSL libc, the result is that `libstdc++` can not use GNU specific features - MUSL doesnt have them. So it falls back to just being a standards compliant C++ implementation. It is technically still the glibc - but, the special sauce is removed. On systems that use the GNU library as default, the resulting `libstdc++` has a few tiny extras as part of the GNU "by-spec". As typical of GNU, they have their little extras now and then, here and there. :) (Look at `gnu++11`, for instance.)

> [...] GCC 5 or 6, and _GLIBCXX_USE_CXX_ABI=1 if building with GCC 7

My version:
```
# gcc --version
gcc (OpenWrt GCC 11.2.0) 11.2.0
Copyright (C) 2021 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

But you already know why. MUSL + GNU C++ library = No GNU extras... mostly. I am sure this is independent of the GCC version. I... think.

> and/or, adjust the mconfig.Linux.sh shell script which is used for configuration on Linux. 

Completely overlooked this one! Right; within a shell script, this is super easy.

```
root@FriendlyWrt ~# cat /etc/os-release
NAME="OpenWrt"
VERSION="22.03.2"
ID="openwrt"
ID_LIKE="lede openwrt"
PRETTY_NAME="OpenWrt 22.03.2"
VERSION_ID="22.03.2"
HOME_URL="https://openwrt.org/"
BUG_URL="https://bugs.openwrt.org/"
SUPPORT_URL="https://forum.openwrt.org/"
BUILD_ID="r19803-9a599fee93"
OPENWRT_BOARD="rockchip/armv8"
OPENWRT_ARCH="aarch64_generic"
OPENWRT_TAINTS="busybox"
OPENWRT_DEVICE_MANUFACTURER="OpenWrt"
OPENWRT_DEVICE_MANUFACTURER_URL="https://openwrt.org/"
OPENWRT_DEVICE_PRODUCT="Generic"
OPENWRT_DEVICE_REVISION="v0"
OPENWRT_RELEASE="OpenWrt 22.03.2 r19803-9a599fee93"
```

I'll work with the shell script then. If the file exists, and the name returns `OpenWrt`, the rest is very easy to guess. Thanks for the reminder!

**Reply by davmac314** [reply]
(Sorry, I had the wrong name in one location above. It's `_GLIBCXX_USE_CXX11_ABI` not `_GLIBCXX_USE_CXX_ABI`. The latter does absolutely nothing).

> Not quite. :) The system's standard C library - I'll just use the term "libc" - is also the basis for the C++ standard library.

Yes, but:

> The GNU C Library - glibc - also has a deriviation.

I'm not sure what you mean by this. I think we are confusing each other by using different terminology. To be clear:

- When I say "Glibc" I mean the GNU C library (specifically, just the C library, not the C++-specific library parts which are provided by libstdc++).
- libstdc++ is the C++ standard library implementation that is normally built as part of GCC.
- Nothing in Glibc is affected by the `_GLIBCXX_USE_CXX11_ABI` macro. The macro only affects libstdc++.
- The _GLIBCXX_USE_CXX11_ABI macro controls  whether to use the C++11 (and newer) ABI in libstdc++, for C++-specific library components (such as `std::string`, exceptions, etc, that are _not_ part of the C library). If set to 0, it forces use of the older (C++03/C++98 and prior) ABI. This is for the C++-specific parts of the ABI,  completely independent from the underlying C library (Glibc, Musl, or anything else). It does not affect availability of any features, just implementation details and inheritance hierarchy for some specific classes.

> When you build GCC, look for a folder of `GLIBCXX_$version`.

My build of GCC doesn't have any such folder... nothing that starts with `GLIBCXX` at all.

> 
> > [...] GCC 5 or 6, and _GLIBCXX_USE_CXX_ABI=1 if building with GCC 7
> 
> My version:
> 
> ```
> # gcc --version
> gcc (OpenWrt GCC 11.2.0) 11.2.0
> Copyright (C) 2021 Free Software Foundation, Inc.
> This is free software; see the source for copying conditions.  There is NO
> warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
> ```

Then you should not need to, and should not, set `_GLIBCXX_USE_CXX11_ABI=0`.

You can verify this using the test case from the relevant GCC bug (https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66145):

```
#include <iostream>
#include <fstream>
using namespace std;
int main ()
{
  std::ifstream pstats;
  pstats.exceptions(ifstream::failbit | ifstream::badbit | ifstream::eofbit);
  try {
      printf("\n Opening file : /proc/0/stat ");
      pstats.open("/proc/0/stat");
  }
  catch(ifstream::failure e)
    {
      printf("\n!!Caught ifstream exception!!\n");
      if(pstats.is_open()) {
          pstats.close();
      }
    }
  return 0;
}
```

If you compile that program (eg `g++ test.cc`), and run it, it should print `!!Caught ifstream exception!!`. If it does, there is no need to set _GLIBCXX_USE_CXX11_ABI to anything in particular when compiling Dinit.

If you run that program and it fails due to an "uncaught exception", with a message something like the following:
```
terminate called after throwing an instance of 'std::ios_base::failure[abi:cxx11]'
  what():  basic_ios::clear: iostream error
Aborted (core dumped)
```
... then you may need to set `_GLIBCXX_USE_CXX11_ABI` to either `0` or `1` (whichever causes it to work correctly).

> 
> But you already know why. MUSL + GNU C++ library = No GNU extras... mostly. I am sure this is independent of the GCC version. I... think.

To be clear, `_GLIBCXX_USE_CXX11_ABI` is not about enabling GNU extras. It is about the choice of ABI to use for a select number of C++ classes. The default, and whether changing it is necessary, is dependent on the GCC version (or specifically, on the libstdc++ version, bearing in mind that libstdc++ is normally considered part of GCC).
 
> I'll work with the shell script then. If the file exists, and the name returns `OpenWrt`, the rest is very easy to guess. Thanks for the reminder!

Great!
